### DISCUSSION TITLE
libcurl POST via proxy with NTLM aborts connection

### DISCUSSION THREAD

**Discussion by genotrance** [curl/curl#15700] [Q&A] [answered]
Attempting an HTTP POST to a local httpbin instance through an NTLM authenticating HTTP proxy. libcurl is not being used to do the actual authentication, it is simply passing on relevant headers to the upstream proxy.

```
NTLM client => libcurl => auth proxy => httpbin
```

Relevant logs:
```
New curl instance
POST http://127.0.0.1:46029/post using HTTP/1.1
Using CAINFO from /tmp/py312/lib/python3.12/site-packages/mcurl/cacert.pem
Path = http://127.0.0.1:46029/post
Proxy = [('127.0.0.1', 3132)]
Skipping proxy authentication

Using easy interface
Curl info: Trying 127.0.0.1:3132...
Curl info: Connected to 127.0.0.1 (127.0.0.1) port 3132
Sent header => POST http://127.0.0.1:46029/post HTTP/1.1
Sent header => Host: 127.0.0.1:46029
Sent header => User-Agent: mcurl v0.10.0
Sent header => Proxy-Authorization: NTLM sanitized len(45)
Sent header => Accept: */*
Sent header => Proxy-Connection: Keep-Alive
Sent header => Content-Length: 0
Sent header => Content-Type: application/x-www-form-urlencoded
Received header <= HTTP/1.1 407 Proxy authentication required
Received header <= Server: BaseHTTP/0.6 Python/3.12.7
Received header <= Date: Thu, 05 Dec 2024 21:02:49 GMT
Received header <= Proxy-Authenticate: NTLM sanitized len(213)
Received header <= Proxy-Connection: Keep-Alive
Received header <= Content-Type: text/html;charset=utf-8
Received header <= Content-Length: 413
Curl info: HTTP error before end of send, stop sending
Curl info: abort upload
Curl info: shutting down connection #0
```

libcurl sends the POST with headers and `Content-Length: 0` as expected. The upstream proxy responds with a challenge. However, libcurl then shuts down the connection for some reason. The response to the NTLM challenge is sent via a new curl instance to a new thread on the authenticating proxy and fails since the NTLM context is lost.

I don't understand why libcurl is shutting down the connection here. Appreciate any guidance around this. 

I haven't seen this issue with older versions of libcurl though that was tested using libcurl via Python [ctypes](https://pypi.org/project/px-proxy/0.9.2/). I am still working on trying out older versions with [pymcurl](https://pypi.org/project/pymcurl/). I will also be trying out newer versions though it will take time to recompile libcurl, pymcurl and re-test. Hoping for some insight in the interim. Thanks in advance.

Using libcurl on Linux, though this issue is reproducible on Windows and Mac as well.
```
libcurl/8.9.1 OpenSSL/3.0.8 zlib/1.3.1 nghttp2/1.64.0
CURL_VERSION_SSL: True
CURL_VERSION_SSPI: False
CURL_VERSION_SPNEGO: True
CURL_VERSION_GSSAPI: True
CURL_VERSION_GSSNEGOTIATE: False
CURL_VERSION_KERBEROS5: True
CURL_VERSION_NTLM: True
CURL_VERSION_NTLM_WB: False
Host: x86_64-pc-linux-musl
```

**Comment by dfandrich** [comment]
What options are you setting?

**Reply by genotrance** [reply]
I'm using the following libcurl options:
- CURLOPT_PROXY = 127.0.0.1
- CURLOPT_PROXYPORT = 3132
- CURLOPT_NOPROXY = NULL
- CURLOPT_CONNECTTIMEOUT = 60
- CURLOPT_CAINFO = cacert.pem
- CURLOPT_POST = True
- CURLOPT_URL = http://127.0.0.1:46029/post
- CURLOPT_HTTP_VERSION = CURL_HTTP_VERSION_1_1
- CURLOPT_DEBUGFUNCTION = wa_callback
- CURLOPT_VERBOSE = True
- CURLOPT_HTTP_TRANSFER_DECODING = False
- CURLOPT_USERAGENT = mcurl v0.10.0
- CURLOPT_READFUNCTION = read_callback
- CURLOPT_READDATA = easyhash
- CURLOPT_WRITEFUNCTION = write_callback
- CURLOPT_WRITEDATA = easyhash
- CURLOPT_HEADERFUNCTION = header_callback
- CURLOPT_HEADERDATA = easyhash

Headers are as per the log above.

The Python [code](https://github.com/genotrance/mcurl/blob/71277e25ee60d1b89afb5c92ffab746182bf9e9e/mcurl/__init__.py#L372) is here if needed for reference.

**Comment by dfandrich** [comment]
I don't see any options to do with authentication in there.

**Reply by genotrance** [reply]
I am not using libcurl to do the authentication, the headers given to libcurl contains the NTLM auth info (`Proxy-Authorization`) as provided by the client. libcurl is simply forwarding the headers to the proxy and getting a response back and shutting down the connection.

**Reply by genotrance** [reply]
Also want to add some stuff I missed about the headers which are set using `curl_slist_append() + CURLOPT_HTTPHEADER`, [code](https://github.com/genotrance/mcurl/blob/71277e25ee60d1b89afb5c92ffab746182bf9e9e/mcurl/__init__.py#L553) is here.

- Setting `Expect:` header to blank
- CURLOPT_POSTFIELDSIZE = Content-Length provided by client

**Comment by dfandrich** [comment]
I see what you're doing now. It seems libcurl is receiving the HTTP response before it has sent the 0 bytes of payload and therefore "aborts" the (zero-length) upload then closes the connection to prevent the connection from getting out of sync. Maybe this is an edge case related to the upload being zero-length, or maybe the local proxy responds so quickly curl doesn't get time to switch from sending the headers to sending the (nonexistent) data and mark the upload as complete. Enabling verbose trace options might provide some clues.

**Reply by genotrance** [reply]
The upstream proxy is checking and responding without waiting for the data, it might be possible that this is a timing issue. I'll put in a delay in the upstream proxy to see if it helps.

As far as verbose [tracing](https://everything.curl.dev/usingcurl/verbose/trace.html) is concerned, `CURLOPT_VERBOSE` is already enabled and the logs above are from [CURLOPT_DEBUGFUNCTION](https://github.com/genotrance/mcurl/blob/71277e25ee60d1b89afb5c92ffab746182bf9e9e/mcurl/__init__.py#L169) but only includes `CURLINFO_TEXT`, `CURLINFO_HEADER_IN` and `CURLINFO_HEADER_OUT`. Do you need the output from `CURLINFO_DATA_IN` and `CURLINFO_DATA_OUT` as well?

**Comment by dfandrich** [comment]
I mean enable curl_global_trace()

**Comment by bagder** [comment, marked as answer]
`CURLOPT_KEEP_SENDING_ON_ERROR` maybe ?

**Reply by genotrance** [reply]
Brilliant! It was added for my exact issue, what are the odds! Thank you @bagder, @dfandrich and @mkauf!

Related:
- #904 
- https://curl.se/mail/lib-2016-06/0061.html
